Use platform independant isRelative.

kolmodin opened the pull request:
We can't use the one from System.FilePath since we don't know if we're
dealing with Windows or Posix paths. See comments in code for more
detail.

This was discovered while working on haskell/hackage-server#533.

kolmodin commented:
Ping @dcoutts

23Skidoo commented on Cabal/Distribution/PackageDescription/Check.hs:
```diff
@@ -627,7 +626,7 @@ checkSourceRepos pkg =
         ++ "or release of the package."
 
-  , check (maybe False System.FilePath.isAbsolute (repoSubdir repo)) $
+  , check (maybe False isAbsolute (repoSubdir repo)) $
```
This is not the only place we use `isAbsolute`/`isRelative`. Do we need to change every occurrence?

23Skidoo commented on Cabal/Distribution/PackageDescription/Check.hs:
```diff
@@ -1904,3 +1903,25 @@ fileExtensionSupportedLanguage path =
     isHaskell = extension `elem` [".hs", ".lhs"]
     isC       = isJust (filenameCDialect extension)
+
+-- Don't use System.FilePath.isRelative. The System.FilePath exists in two
```
If the answer to my comment above is positive, I suggest moving these two functions to `Distribution.Simple.Utils`.

ezyang commented:
Yeah, checking if there are other places in the code that would benefit from this would be appreciated, but if you don't want to we can just merge this.

kolmodin commented:
There are other uses of `isRelative` and `isAbsolute` throughout Cabal, but most of them needs to be platform independant. The issue with the `.Check` module is that it may run on packages that are made for another platform, like hackage checking windows packages.

This is not the case for most other parts of the code which has to do with `configure`, `haddock`, `register`, `install`.

`sdist` has one case where looks for include files in from `include-dirs`. Currently, it looks like if you try to include files with absolute paths of another platform than what you're using yourself, sdist will fail.

I can fix that with another patch this weekend.

dcoutts commented on Cabal/Distribution/PackageDescription/Check.hs:
```diff
@@ -1904,3 +1903,25 @@ fileExtensionSupportedLanguage path =
+-- System.FilePath.Posix.isAbsolute \"C:\\hello\" == False
+-- System.FilePath.Windows.isAbsolute \"/hello\" == False
+-- This means that we would treat paths that start with \"/\" to be absolute.
+-- On Posix they are indeed absolute, while on Windows they are not.
```
Ugg, I see. This is more nasty than I thought.

kolmodin commented:
Moved `isAbsolute` and `isRelative` to the `Utils` module. Also changed all the code to make use of those versions, not the ones from `System.FilePath`.

23Skidoo commented:
Merged, thanks!

dcoutts commented:
@23Skidoo @kolmodin Wait, hold on, I think we've made a mistake here. The "portable" ones can only be used for the "Check" modules, not everywhere. The "check" module is used on the server and it also makes sense on clients to check in the context of not knowing if the target system will be windows or unix.

Everywhere else we should continue to use the normal System.FilePath ones! Can we revert and use the earlier version of the patch where it's not in the Utils module, just in the Check module.

I'm pretty sure @ezyang's comment was meant only within the context of the Check module, not the whole codebase :-)

kolmodin commented:
@dcoutts Ok, so we can let the `.Check` module and `cabal sdist` command use the platform independent versions since they might run on code which you're not building/installing.

I can make a PR to let `configure`, `haddock`, `register`, `install` use the normal System.FilePath as they all mean that you're using the package on the intended platform.
SG?

ezyang commented:
So, is there a note anywhere saying when you should use our ones and when you shouldn't? Seems tricky and worth warning people about.

dcoutts commented:
Should add a note, but the principle is if you're doing portability checks then use the special ones, but in all normal cases use the System.FilePath (ie native) ones.

23Skidoo commented:
@dcoutts 
My thinking in merging this was that the probability of us encountering relative paths that start with `\\` or `C:\` on *nix is very low (even though they're legal), so instead of adding special cases for sdist and check it makes sense to just use a single implementation everywhere.
